### PR TITLE
(FM-7903) - Implement Puppet Strings

### PR DESCRIPTION
**Pull request by david22swan**

### CODE REVIEW

File: REFERENCE.md [hunk 1]
```diff
@@ -0,0 +1,1404 @@
...
+Set to true to match tcp fragments (requires type to be set to tcp)
+
+##### `recent`
+
+Valid values: set, update, rcheck, remove
+
+Enable the recent module. Takes as an argument one of set, update,
+rcheck or remove. For example:
+
+  # If anyone's appeared on the 'badguy' blacklist within
```
Comment (eimlav) [inline REFERENCE.md]: Formatting is broken for the following examples due to the comments.

File: REFERENCE.md [hunk 2]
```diff
@@ -0,0 +1,1404 @@
...
+This is matched against the output of `iptables-save`.
+
+This can be a single regex, or an array of them.
+To support flags, use the ruby inline flag mechanism.
+Meaning a regex such as
+  /foo/i
+can be written as
+  '(?i)foo' or '(?i:foo)'
+
+Full example:
```
Comment (eimlav) [inline REFERENCE.md]: Formatting for example is broken.

File: README.markdown [hunk 3]
```diff
@@ -365,582 +364,6 @@ Or
...
-
-Enables you to manage rule chains for firewalls.
-
-Currently this type supports only iptables, ip6tables, and ebtables on Linux. It also provides support for setting the default policy on chains and tables that allow it.
-
-**Autorequires**: If Puppet is managing the iptables or iptables-persistent packages, and the provider is iptables_chain, the firewall resource will autorequire those packages to ensure that any required binaries are installed.
-
-#### Providers
-
-`iptables_chain` is the only provider that supports firewallchain.
```
Comment (eimlav) [inline README.markdown]: Missing from REFERENCE.md. May need to be included as part of the description for firewallchain.

File: README.markdown [hunk 4]
```diff
@@ -365,582 +364,6 @@ Or
...
-
-#### Features
```
Comment (eimlav) [inline README.markdown]: Also missing from REFERENCE. May need to be included as part of the description for firewallchain.

File: REFERENCE.md [hunk 5]
```diff
@@ -0,0 +1,1404 @@
...
+
+For example:
+
+    1-1024
+
+This would cover ports 1 to 1024.
+
+##### `port`
+
+DEPRECATED
```
Comment (eimlav) [inline REFERENCE.md]: This deprecation notice should be more explicit. Might be worth using an `@note`.

File: REFERENCE.md [hunk 6]
```diff
@@ -0,0 +1,1404 @@
...
+
+For example:
+
+    1-1024
+
+This would cover ports 1 to 1024.
+
+##### `dst_type`
+
+Valid values: [:UNSPEC, :UNICAST, :LOCAL, :BROADCAST, :ANYCAST, :MULTICAST,
```
Comment (eimlav) [inline REFERENCE.md]: Formatting broken here.

File: REFERENCE.md [hunk 7]
```diff
@@ -0,0 +1,1404 @@
...
+
+    dst_type => ['LOCAL --limit-iface-in']
+
+It can also be negated using '!':
+
+    dst_type => ['! LOCAL']
+
+Will accept a single element or an array.
+
+##### `src_type`
```
Comment (eimlav) [inline REFERENCE.md]: Formatting also broken here.

File: REFERENCE.md [hunk 8]
```diff
@@ -0,0 +1,1404 @@
...
+
+It can also be negated using '!':
+
+    src_type => ['! LOCAL']
+
+Will accept a single element or an array.
+
+##### `proto`
+
+Valid values: [:ip, :tcp, :udp, :icmp, :"ipv6-icmp", :esp, :ah, :vrrp, :igmp, :ipencap, :ipv4, :ipv6, :ospf, :gre, :cbt, :sctp, :pim, :all].map { |proto|
```
Comment (eimlav) [inline REFERENCE.md]: Formatting broken here.

File: REFERENCE.md [hunk 9]
```diff
@@ -0,0 +1,1404 @@
...
+      [proto, "! #{proto}".to_sym]
+    }.flatten
+
+The specific protocol to match for this rule. By default this is
```
Comment (eimlav) [inline REFERENCE.md]: This should be removed as default is already specified.

File: REFERENCE.md [hunk 10]
```diff
@@ -0,0 +1,1404 @@
...
+
+Only match on the given days of the month. Possible values are 1 to 31.
+Note that specifying 31 will of course not match on months which do not have a 31st day;
+the same goes for 28- or 29-day February.
+
+##### `week_days`
+
+Valid values: Mon, Tue, Wed, Thu, Fri, Sat, Sun
+
+Only match on the given weekdays. Possible values are Mon, Tue, Wed, Thu, Fri, Sat, Sun.
```
Comment (eimlav) [inline REFERENCE.md]: The `possible values` section should be removed as it is already stated in `Valid values`.

File: README.markdown [hunk 11]
```diff
@@ -365,582 +364,6 @@ Or
...
-
-  **Note** This `purge` is purging unmanaged rules in a firewall chain, not unmanaged firewall chains. To purge unmanaged firewall chains, use the following instead.
-
-  ~~~ puppet
-  resources { 'firewallchain':
-    purge => true,
-  }
-  ~~~
-
-### Fact: ip6tables_version
```
[the remark this answers is not in the document]
Comment (david22swan) [reply README.markdown]: Have added back

File: REFERENCE.md [hunk 12]
```diff
@@ -0,0 +1,1404 @@
+# Reference
+<!-- DO NOT EDIT: This document was generated by Puppet Strings -->
+
+## Table of Contents
+
+**Classes**
+
+_Public Classes_
+
+* [`firewall`](#firewall): Main class, includes all other classes
```
[the remark this answers is not in the document]
Comment (david22swan) [reply REFERENCE.md]: Swapped it to the old summary

File: REFERENCE.md [hunk 13]
```diff
@@ -0,0 +1,1404 @@
...
+* [`firewall::linux`](#firewalllinux): Main class, includes all other classes
```
[the remark this answers is not in the document]
Comment (david22swan) [reply REFERENCE.md]: set to private

File: REFERENCE.md [hunk 14]
```diff
@@ -0,0 +1,1404 @@
...
+
+##### `ebtables_manage`
+
+Data type: `Boolean`
+
+Controls whether puppet manages the ebtables package or not. If managed, the package will use the value of pkg_ensure.
+
+Default value: `false`
+
+### firewall::linux
```
[the remark this answers is not in the document]
Comment (david22swan) [reply REFERENCE.md]: is private

File: REFERENCE.md [hunk 15]
```diff
@@ -0,0 +1,1404 @@
...
+
+##### `package_ensure`
+
+Data type: `Enum['present', 'latest']`
+
+Controls the state of the iptables package on your system. Valid options: 'present' or 'latest'. Defaults to 'latest'.
+
+Default value: $::firewall::params::package_ensure
+
+### firewall::linux::debian
```
[the remark this answers is not in the document]
Comment (david22swan) [reply REFERENCE.md]: same

File: REFERENCE.md [hunk 16]
```diff
@@ -0,0 +1,1404 @@
...
+
+##### `package_ensure`
+
+Data type: `Enum['present', 'latest']`
+
+Controls the state of the iptables package on your system. Valid options: 'present' or 'latest'. Defaults to 'latest'.
+
+Default value: $::firewall::params::package_ensure
+
+### firewall::linux::redhat
```
[the remark this answers is not in the document]
Comment (david22swan) [reply REFERENCE.md]: private now

File: REFERENCE.md (hunk 9 above)
Comment (david22swan) [reply REFERENCE.md]: Removed

File: REFERENCE.md (hunk 10 above)
Comment (david22swan) [reply REFERENCE.md]: Removed

File: REFERENCE.md (hunk 5 above)
Comment (david22swan) [reply REFERENCE.md]: There is no note for the types

File: README.markdown (hunk 4 above)
Comment (david22swan) [reply README.markdown]: added to the summary

File: REFERENCE.md [hunk 17]
```diff
@@ -0,0 +1,1404 @@
...
+
+##### `package_ensure`
+
+Data type: `Enum['present', 'latest']`
+
+Controls the state of the iptables package on your system. Valid options: 'present' or 'latest'. Defaults to 'latest'.
+
+Default value: $::firewall::params::package_ensure
+
+### firewall::linux::gentoo
```
[the remark this answers is not in the document]
Comment (david22swan) [reply REFERENCE.md]: privcate

File: README.markdown [hunk 18]
```diff
@@ -365,582 +364,6 @@ Or
...
-    * Default for `kernel` == `linux`.
-    * Supported features: `address_type`, `clusterip`, `connection_limiting`, `dnat`, `icmp_match`, `interface_match`, `iprange`, `ipsec_dir`, `ipsec_policy`, `ipset`, `iptables`, `isfragment`, `length`, `log_level`, `log_prefix`, `log_uid`, `mark`, `mask`, `mss`, `netmap`, `nflog_group`, `nflog_prefix`, `nflog_range`, `nflog_threshold`, `owner`, `pkttype`, `queue_bypass`, `queue_num`, `rate_limiting`, `recent_limiting`, `reject_type`, `snat`, `socket`, `state_match`, `string_matching`, `tcp_flags`, `bpf`.
-
-**Autorequires:**
-
-If Puppet is managing the iptables or ip6tables chains specified in the `chain` or `jump` parameters, the firewall resource will autorequire those firewallchain resources.
-
-If Puppet is managing the iptables or iptables-persistent packages, and the provider is iptables or ip6tables, the firewall resource will autorequire those packages to ensure that any required binaries are installed.
-
-#### Features
```
[the remark this answers is not in the document]
Comment (david22swan) [reply README.markdown]: added to summary

File: README.markdown (hunk 3 above)
Comment (david22swan) [reply README.markdown]: added

File: README.markdown [hunk 19]
```diff
@@ -365,582 +364,6 @@ Or
...
-
-#### package_name
-
-Specify the platform-specific package(s) to install. Defaults defined in `firewall::params`.
-
-### Type: firewall
-
-This type enables you to manage firewall rules within Puppet.
-
-#### Providers
```
[the remark this answers is not in the document]
Comment (david22swan) [reply README.markdown]: added

File: REFERENCE.md [hunk 20]
```diff
@@ -0,0 +1,1230 @@
...
+### firewall
+
+Performs the basic setup tasks required for using the firewall resources.
+
+At the moment this takes care of:
+
+iptables-persistent package installation
+Include the firewall class for nodes that need to use the resources in this module:
+
+class { 'firewall': }
```
Comment (eimlav) [inline REFERENCE.md]: An example tag should be used here

File: REFERENCE.md [hunk 21]
```diff
@@ -0,0 +1,1230 @@
...
+
+#### Parameters
+
+The following parameters are available in the `firewall` class.
+
+##### `ensure`
+
+Data type: `Any`
+
+Controls the state of the ipv4 iptables service on your system. Valid options: 'running' or 'stopped'. Defaults to 'running'.
```
Comment (eimlav) [inline REFERENCE.md]: Remove default info.

File: REFERENCE.md [hunk 22]
```diff
@@ -0,0 +1,1230 @@
...
+
+Default value: running
+
+##### `ensure_v6`
+
+Data type: `Any`
+
+Controls the state of the ipv6 iptables service on your system. Valid options: 'running' or 'stopped'. Defaults to 'running'.
```
Comment (eimlav) [inline REFERENCE.md]: Remove default info.

File: REFERENCE.md [hunk 23]
```diff
@@ -0,0 +1,1230 @@
...
+
+Default value: `undef`
+
+##### `pkg_ensure`
+
+Data type: `Any`
+
+Controls the state of the iptables package on your system. Valid options: 'present' or 'latest'. Defaults to 'latest'.
```
Comment (eimlav) [inline REFERENCE.md]: Remove default info.

File: REFERENCE.md [hunk 24]
```diff
@@ -0,0 +1,1230 @@
...
+
+Default value: present
+
+##### `service_name`
+
+Data type: `Any`
+
+Specify the name of the IPv4 iptables service. Defaults defined in firewall::params.
```
Comment (eimlav) [inline REFERENCE.md]: Remove default info.

File: REFERENCE.md [hunk 25]
```diff
@@ -0,0 +1,1230 @@
...
+
+Default value: $::firewall::params::service_name
+
+##### `service_name_v6`
+
+Data type: `Any`
+
+Specify the name of the IPv6 iptables service. Defaults defined in firewall::params.
```
Comment (eimlav) [inline REFERENCE.md]: Remove default info.

File: REFERENCE.md [hunk 26]
```diff
@@ -0,0 +1,1230 @@
...
+
+Default value: $::firewall::params::service_name_v6
+
+##### `package_name`
+
+Data type: `Any`
+
+Specify the platform-specific package(s) to install. Defaults defined in firewall::params.
```
Comment (eimlav) [inline REFERENCE.md]: Remove default info.

File: REFERENCE.md [hunk 27]
```diff
@@ -0,0 +1,1230 @@
...
+
+#### Properties
+
+The following properties are available in the `firewall` type.
+
+##### `ensure`
+
+Valid values: present, absent
+
+Manage the state of this rule. The default action is *present*.
```
Comment (eimlav) [inline REFERENCE.md]: Remove default info.

File: REFERENCE.md (hunk 5 above)
Comment (eimlav) [reply REFERENCE.md]: Maybe just say something like:
`*Note* This property is deprecated`

File: REFERENCE.md [hunk 28]
```diff
@@ -0,0 +1,1230 @@
...
+
+* INPUT
+* FORWARD
+* OUTPUT
+* PREROUTING
+* POSTROUTING
+
+Or you can provide a user-based chain.
+
+The default value is 'INPUT'.
```
Comment (eimlav) [inline REFERENCE.md]: Remove default info.

File: REFERENCE.md [hunk 29]
```diff
@@ -0,0 +1,1230 @@
...
+
+Table to use. Can be one of:
+
+* nat
+* mangle
+* filter
+* raw
+* rawpost
+
+By default the setting is 'filter'.
```
Comment (eimlav) [inline REFERENCE.md]: Remove default info.

File: REFERENCE.md [hunk 30]
```diff
@@ -0,0 +1,1230 @@
...
+
+Valid values: set, update, rcheck, remove
+
+Enable the recent module. Takes as an argument one of set, update,
+rcheck or remove. For example:
+
+  If anyone's appeared on the 'badguy' blacklist within
+  the last 60 seconds, drop their traffic, and update the timestamp.
+
+  `firewall { '100 Drop badguy traffic':
```
Comment (eimlav) [inline REFERENCE.md]: Newlines are not rendering properly. Try using ``` above and below examples. This should preserve comments too so use `#` for those.
```
firewall { '100 Drop badguy traffic':
...
}
```

File: REFERENCE.md [hunk 31]
```diff
@@ -0,0 +1,1230 @@
...
+
+Valid values: in, out
+
+Sets the ipsec policy direction
+
+##### `stat_mode`
+
+Valid values: nth, random
+
+Set the matching mode for statistic matching. Supported modes are `random` and `nth`.
```
Comment (eimlav) [inline REFERENCE.md]: Remove supported modes info.

File: REFERENCE.md [hunk 32]
```diff
@@ -0,0 +1,1230 @@
...
+
+Only match on the given days of the month. Possible values are 1 to 31.
+Note that specifying 31 will of course not match on months which do not have a 31st day;
+the same goes for 28- or 29-day February.
+
+##### `week_days`
+
+Valid values: Mon, Tue, Wed, Thu, Fri, Sat, Sun
+
+Only match on the given weekdays..
```
Comment (eimlav) [inline REFERENCE.md]: Extra full stop here.

File: REFERENCE.md [hunk 33]
```diff
@@ -0,0 +1,1230 @@
...
+
+Used with the CLUSTERIP jump target.
+Create a new ClusterIP. You always have to set this on the first rule for a given ClusterIP.
+
+##### `clusterip_hashmode`
+
+Valid values: sourceip, sourceip-sourceport, sourceip-sourceport-destport
+
+Used with the CLUSTERIP jump target.
+Specify the hashing mode. Valid values: sourceip, sourceip-sourceport, sourceip-sourceport-destport.
```
Comment (eimlav) [inline REFERENCE.md]: Remove default info.

File: REFERENCE.md (hunk 2 above)
Comment (eimlav) [reply REFERENCE.md]: Fix formatting here using ``` on the line before and after the example.

File: REFERENCE.md (hunk 5 above)
Comment (david22swan) [reply REFERENCE.md]: made it '*note* This property has been DEPRECATED'

File: REFERENCE.md (hunk 30 above)
Comment (david22swan) [reply REFERENCE.md]: worked

File: REFERENCE.md (hunk 6 above)
Comment (david22swan) [reply REFERENCE.md]: This is taken from code, not documentation. Not something I can format.
```
newproperty(:proto) do
    desc <<-PUPPETCODE
      The specific protocol to match for this rule.
    PUPPETCODE

    newvalues(*[:ip, :tcp, :udp, :icmp, :"ipv6-icmp", :esp, :ah, :vrrp, :igmp, :ipencap, :ipv4, :ipv6, :ospf, :gre, :cbt, :sctp, :pim, :all].map { |proto|
      [proto, "! #{proto}".to_sym]
    }.flatten)
    defaultto 'tcp'
  end
```

File: README.markdown [hunk 34]
```diff
@@ -367,568 +367,14 @@ Or
...
 
-Classes:
-
-* [firewall](#class-firewall)
-
-Types:
-
-* [firewall](#type-firewall)
-* [firewallchain](#type-firewallchain)
+Please refer to the REFERENCE.md for more information on the various Classes and Types of this module. Information regarding it's Facts is shown below.
```
Comment (clairecadman) [inline README.markdown]: Suggested change: "For information on the classes and types, see the REFERENCE.md. For information on the facts, see below."

Comment (david22swan) [reply README.markdown]: Have updated the writing to what you suggested. Have also added a link the the REFERENCE.md

Comment (clairecadman) [reply README.markdown]: Great!

**Pull request merged**